File: python/huffman.py
from collections import defaultdict, namedtuple
from typing import Optional, DefaultDict, List, Callable, Dict, Tuple
import heapq

NULLCHAR = "\x07f"

Node = namedtuple('Node', ['freq', 'c', 'left', 'right'])

def leaf(f: float, c: str) -> Node:
    return Node(f, c, None, None)

def compute_code_tree(s: str) -> Optional[Node]:
    count : DefaultDict[str,int] = defaultdict(int)
    n = 0
    for c in s:
        count[c] += 1
        n += 1

    q : List[Node] = []
    for k, v in count.items():
        heapq.heappush(q, leaf(v / n, k))

    if len(q) == 0:
        return None

    while len(q) > 1:
        r = heapq.heappop(q)
        l = heapq.heappop(q)
        t = Node(l.freq + r.freq, NULLCHAR, l, r)
        heapq.heappush(q, t)

    return heapq.heappop(q)


def dfs(t : Optional[Node],
        f : Callable[[Optional[Node], List[str]], None],
        p : List[str]):
    if t is None:
        return
    else:
        dfs(t.left, f, p + ['0'])
        f(t, p)
        dfs(t.right, f, p + ['1'])


def compute_code_table(t : Optional[Node]) -> Dict[str,str]:
    r = {}

    def visit(n, p):
        if n.c != NULLCHAR:
            r[n.c] = "".join(p)

    dfs(t, visit, [])
    return r


def gen_huffman(s: str) -> Tuple[Dict[str,str], Optional[Node]]:
    decoder = compute_code_tree(s)
    encoder = compute_code_table(decoder)
    return encoder, decoder


def huffman_encode(encoder: Dict[str,str], s: str) -> str:
    return "".join(encoder[c] for c in s)


def decode_char(t: Node, bs: str) -> Tuple[str, str]:
    if t.c != NULLCHAR:
        return t.c, bs
    else:
        if bs[0] == '0':
            return decode_char(t.left, bs[1:])
        else:
            return decode_char(t.right, bs[1:])


def huffman_decode(decoder: Optional[Node], bs: str) -> str:
    assert decoder is not None
    r = []
    while bs:
        c, bs = decode_char(decoder, bs)
        r.append(c)
    return "".join(r)


test_data = "foo bar baz bee"
encoder, decoder = gen_huffman(test_data)

print("data:", test_data)
print("encoder:", encoder)
encoded_data = huffman_encode(encoder, test_data)
print("encoded data:", encoded_data, "bits:", len(encoded_data))
decoded_data = huffman_decode(decoder, encoded_data)
print("decoded data:", decoded_data, "bits:", len(test_data) * 8)
assert test_data == decoded_data

File: julia/huffman.jl
import DataStructures: BinaryMinHeap, nil, cons, LinkedList, Nil

const NULLCHAR = '\x7f'

mutable struct Node
    freq::Int
    c::Char
    left::Union{Node,Nothing}
    right::Union{Node,Nothing}
end

Base.isless(a::Node, b::Node) = a.freq < b.freq 

function compute_code_tree(s::String)::Union{Node,Nothing}
    d = Dict{Char,Int}()
    for c in s
        d[c] = get(d, c, 0) + 1
    end
    nodes = Vector{Node}()
    for (ch,fr) in d
        push!(nodes, Node(fr, ch, nothing, nothing))
    end

    h = BinaryMinHeap(nodes)
    while length(h) > 1
        r = pop!(h)
        l = pop!(h)
        push!(h, Node(l.freq + r.freq, NULLCHAR, l, r))
    end

    if length(h) == 0
        nothing
    else
        pop!(h)
    end
end

function dfs(t::Union{Node,Nothing}, f::Function, p::LinkedList{Char})
    if t == nothing
        return
    else
        dfs(t.left, f, cons('0', p))
        f(t, p)
        dfs(t.right, f, cons('1', p))
    end
end

function compute_code_table(tree::Union{Node,Nothing})::Dict{Char,String}
    r = Dict{Char,String}()

    visit = (n::Node, p::LinkedList{Char}) -> 
        if n.c != NULLCHAR
            r[n.c] = join(reverse(p), "")
        end

    dfs(tree, visit, Nil{Char}())
    r
end

huffman_encode(encoder, s) = join([encoder[c] for c in s], "")

function huffman_decode_char(tree::Node, bs::String)
    if tree.c != NULLCHAR
        tree.c, bs
    else
        b,bs = bs[1],bs[2:end]
        next = if b == '0' tree.left else tree.right end
        huffman_decode_char(next, bs)
    end
end

function huffman_decode(decoder::Union{Node,Nothing}, bs::String)
    if decoder == nothing
        return ""
    end
    r = Vector{Char}()
    while length(bs) > 0
        c, bs = huffman_decode_char(decoder, bs)
        push!(r, c)
    end
    join(r, "")
end

function main()
    s = "foo bar baz bee"
    decoder = compute_code_tree(s)
    encoder = compute_code_table(decoder)
    println("Encoder:", encoder)
    ed = huffman_encode(encoder, s)
    println("Encoded data: ", ed, " Len: ", length(ed))
    dd = huffman_decode(decoder, ed)
    println("Decoded data: ", dd, " Len: ", length(dd))
end

main()

